sycl/doc/extensions/proposed/sycl_ext_oneapi_address_space_cast.asciidoc

```diff
@@ -0,0 +1,208 @@
+dynamic_address_space_cast(ElementType* ptr);
+----
+_Preconditions_: The memory at `ptr` is accessible to the calling work-item.
+
+_Returns_: A `multi_ptr` with the specified address space and decoration that
+points to the same object as `ptr` if `ptr` points to an object allocated in
+the address space designated by `Space`, and `nullptr` otherwise.
+
+_Diagnostics_: Implementations using address space deduction must issue a
+diagnostic if `ptr` does not point to an object allocated in the address space
+designated by `Space`.
+
+[NOTE]
+====
+The precondition prevents `dynamic_address_space_cast` from being used to
+reason about the address space of pointers originating from another work-item
+(in the case of `private` pointers) or another work-group (in the case of
+`local` pointers). Such pointers could not be dereferenced by the calling
+work-item, and it is thus unclear that being able to reason about the address
+space would be useful. Limiting the use of `dynamic_address_space_cast` to
+accessible pointers is expected to result in simpler and faster
+implementations.
+====
+
+
+== Implementation notes
+
+For SPIR-V backends, `static_address_space_cast` corresponds to
+`OpGenericCastToPtr`. `dynamic_address_space_cast` _may_ correspond to
+`OpGenericCastToPtrExplicit` -- there is currently some ambiguity regarding
+exactly how `OpGenericCastToPtrExplicit` is expected to behave, because the
+SPIR-V specification does not explain what it means for a cast to "fail".
+Since this extension is only experimental, we can likely implement
+`dynamic_address_space_cast` using `OpGenericCastToPtrExplicit` while we
+seek to clarify the SPIR-V specification.
+
+Generally speaking, it is expected that a `static_address_space_cast` can
+simply attach new decoration(s) to the raw pointer (or do nothing), while
+a `dynamic_address_space_cast` will have to inspect the address of the
+raw pointer to determine which region of memory it points to.
+
+An implementation for a CPU target could be implemented by keeping track of
+three pieces of information in thread-local storage:
+
+- The base (highest address) of the calling thread's stack.
+- The low bound of the calling work-item's local memory area.
+- The high bound of the calling work-item's local memory area.
+
+A cast to `private_space` succeeds as long as the pointer is within the calling
+thread's stack. A cast to `local_space` succeeds as long as the pointer is
+within the calling work-item's local memory area. A cast to `global_space`
+succeeds as long as the pointer is not within either of the above two address
+ranges.
+
+Implementations for GPU targets may be able to leverage dedicated instructions
+for checking the address space.
+
+
+== Issues
+
+. `static_address_space_cast` and `dynamic_address_space_cast` are long names.
++
+--
+*UNRESOLVED*:
+Dropping one of the words would be less verbose; either `static_address_cast`
+or `static_space_cast` could be considered.
```

Comment:
llvm uses `addrspacecast`, so `static_addrspace_cast` ?

Reply:
@gmlueck, @bashbaug: Before I start implementing this, any opinions on `addrspace_cast`?  I like it, but every other instance of `address_space` in the SYCL 2020 specification is spelled out in full.  (Maybe we could shorten those, too?)

Reply:
I don't feel strongly, but I would probably keep the long version until we decide on some standard shortened version in the rest of the SYCL spec.  FWIW, I don't think the name "addrspace" saves that much because the names of the address spaces themselves are so horribly long (especially when you write out all the namespaces):

```
sycl::dynamic_address_space_cast<sycl::access::address_space::global_space>(ptr)
sycl::dynamic_addrspace_cast<sycl::access::address_space::global_space>(ptr)
```

I think we could eliminate the "access" namespace and somehow shorten the names "global_space", etc.

Reply:
Good point, I hadn't thought about the template argument.

Even keeping `address_space` in full, but following the convention for enums (e,g, `sycl::memory_order_relaxed`) would be quite a bit shorter:

```c++
sycl::dynamic_address_space_cast<sycl::access::address_space::global_space>(ptr)
sycl::dynamic_address_space_cast<sycl::address_space_global>(ptr);
```

Dropping "space" from the cast or shortening to "addrspace" everywhere has a similar effect:
```c++
sycl::dynamic_address_space_cast<sycl::access::address_space::global_space>(ptr)
sycl::dynamic_address_cast<sycl::address_space_global>(ptr);
sycl::dynamic_addrspace_cast<sycl::addrspace_global>(ptr);
```

Of the options, I'm quite fond of the middle one -- everything is a complete word, but it's still manageable.  Dropping "space" from the cast doesn't impact readability, because it's still there in the template argument.

But you've convinced me we shouldn't do any renaming yet, and so I'll leave things as `address_space` for now.

Reply:
Done in b75fe4273b0d01045e680aa33a63b1c552c4f8bc.